clang/lib/Analysis/LifetimeSafety/Origins.cpp

```diff
@@ -7,9 +7,33 @@
 //===----------------------------------------------------------------------===//
 
 #include "clang/Analysis/Analyses/LifetimeSafety/Origins.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/TypeBase.h"
+#include "clang/Analysis/Analyses/LifetimeSafety/LifetimeSafety.h"
+#include "llvm/ADT/StringMap.h"
 
 namespace clang::lifetimes::internal {
 
+static bool isGslPointerType(QualType QT) {
+  if (const auto *RD = QT->getAsCXXRecordDecl()) {
+    // We need to check the template definition for specializations.
+    if (auto *CTSD = dyn_cast<ClassTemplateSpecializationDecl>(RD))
+      return CTSD->getSpecializedTemplate()
+          ->getTemplatedDecl()
+          ->hasAttr<PointerAttr>();
+    return RD->hasAttr<PointerAttr>();
+  }
+  return false;
+}
```

Comment:
FYI: I created a PR https://github.com/llvm/llvm-project/pull/169620 to move this to LifetimeAnnotations.h/.cpp as this also coincides with one of my PRs.

Will let you know if it gets submitted before then we can remove this from here.

Reply:
This has landed. You can remove this after rebasing.